cpp/src/datetime/datetime_ops.cu

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/types.hpp>
+#include <cudf/wrappers/timestamps.hpp>
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/utilities/traits.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <thrust/transform.h>
+#include <thrust/functional.h>
+#include <thrust/execution_policy.h>
+
+namespace cudf {
+namespace datetime {
+namespace detail {
+
+  typedef enum {
+    invalid = 0,
+    year,
+    month,
+    day,
+    weekday,
+    hour,
+    minute,
+    second,
+  } datetime_component;
+
+  template <datetime_component DatetimeComponent>
+  struct extract_component_operator {
+
+    column_device_view column;
+
+    CUDA_HOST_DEVICE_CALLABLE extract_component_operator(column_device_view col) : column(col) {}
+
+    template <typename Element>
```

Comment:
By moving `Element` to a template type of `extract_component_operator`, you could instead add a `static_assert(is_timestamp<Element>(),"")`